File: app/src/main/java/com/dhsdevelopments/potato/editor/UserNameSuggestFilter.java
package com.dhsdevelopments.potato.editor;

import android.widget.Filter;
import com.dhsdevelopments.potato.Log;
import com.dhsdevelopments.potato.userlist.ChannelUsersTracker;

import java.text.Collator;
import java.util.*;

class UserNameSuggestFilter extends Filter
{
    private ChannelUsersTracker usersTracker;
    private UserNameSuggestAdapter userNameSuggestAdapter;

    private UserTrackerListener listener;
    private List<UserSuggestion> users = null;
    private Comparator<UserSuggestion> userSuggestionComparator;

    public UserNameSuggestFilter( ChannelUsersTracker usersTracker, UserNameSuggestAdapter userNameSuggestAdapter ) {
        this.usersTracker = usersTracker;
        this.userNameSuggestAdapter = userNameSuggestAdapter;

        final Collator collator = Collator.getInstance();
        userSuggestionComparator = new Comparator<UserSuggestion>()
        {
            @Override
            public int compare( UserSuggestion o1, UserSuggestion o2 ) {
                return collator.compare( o1.getName(), o2.getName() );
            }
        };

        listener = new UserTrackerListener();
        usersTracker.addUserActivityListener( listener );
    }

    @Override
    protected FilterResults performFiltering( CharSequence text ) {
        if( text == null ) {
            return null;
        }

        if( text.charAt( 0 ) != '@' ) {
            Log.w( "Attempt to filter a string which does not start with @: \"" + text + "\"" );
            return null;
        }

        if( text.length() < 2 ) {
            return null;
        }

        String textLower = text.toString().substring( 1 ).toLowerCase( Locale.getDefault() );
        List<UserSuggestion> result = new ArrayList<>();
        for( UserSuggestion s : getUsers() ) {
            String[] parts = s.getName().toLowerCase( Locale.getDefault() ).split( "\\W+" );
            for( String part : parts ) {
                if( part.startsWith( textLower ) ) {
                    result.add( s );
                    break;
                }
            }
        }

        FilterResults filterResults = new FilterResults();
        filterResults.values = result;
        filterResults.count = result.size();
        return filterResults;
    }

    @SuppressWarnings( "unchecked" )
    @Override
    protected void publishResults( CharSequence charSequence, FilterResults filterResults ) {
        if( filterResults != null ) {
            userNameSuggestAdapter.setSuggestionList( (List<UserSuggestion>)filterResults.values );
        }
    }

    private synchronized List<UserSuggestion> getUsers() {
        if( users == null ) {
            Map<String, ChannelUsersTracker.UserDescriptor> userlist = usersTracker.getUsers();
            users = new ArrayList<>( userlist.size() );
            for( Map.Entry<String, ChannelUsersTracker.UserDescriptor> u : userlist.entrySet() ) {
                users.add( new UserSuggestion( u.getKey(), u.getValue().getName() ) );
            }
            Collections.sort( users, userSuggestionComparator );
        }
        return users;
    }

    private synchronized void clearUserList() {
        users = null;
    }

    public void shutdown() {
        usersTracker.removeUserActivityListener( listener );
    }

    private class UserTrackerListener implements ChannelUsersTracker.UserActivityListener
    {
        @Override
        public void activeUserListSync() {
            clearUserList();
        }

        @Override
        public void userUpdated( String uid ) {
            clearUserList();
        }
    }
}

File: app/src/main/java/com/dhsdevelopments/potato/editor/UserNameSuggestFilter.kt
package com.dhsdevelopments.potato.editor

import android.widget.Filter
import com.dhsdevelopments.potato.Log
import com.dhsdevelopments.potato.userlist.ChannelUsersTracker

import java.text.Collator
import java.util.*

internal class UserNameSuggestFilter(private val usersTracker: ChannelUsersTracker,
                                     private val userNameSuggestAdapter: UserNameSuggestAdapter) : Filter() {

    private val listener: UserTrackerListener
    private var users: List<UserSuggestion>? = null
    private val userSuggestionComparator: Comparator<UserSuggestion>

    init {
        val collator = Collator.getInstance()
        userSuggestionComparator = Comparator<UserSuggestion> { o1, o2 -> collator.compare(o1.name, o2.name) }

        listener = UserTrackerListener()
        usersTracker.addUserActivityListener(listener)
    }

    override fun performFiltering(text: CharSequence?): Filter.FilterResults? {
        if (text == null) {
            return null
        }

        if (text[0] != '@') {
            Log.w("Attempt to filter a string which does not start with @: \"" + text + "\"")
            return null
        }

        if (text.length < 2) {
            return null
        }

        val textLower = text.toString().substring(1).toLowerCase(Locale.getDefault())
        val result = ArrayList<UserSuggestion>()
        for (s in getUsers()) {
            val parts = s.name.toLowerCase(Locale.getDefault()).split("\\W+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (part in parts) {
                if (part.startsWith(textLower)) {
                    result.add(s)
                    break
                }
            }
        }

        val filterResults = Filter.FilterResults()
        filterResults.values = result
        filterResults.count = result.size
        return filterResults
    }

    @SuppressWarnings("unchecked")
    override fun publishResults(charSequence: CharSequence?, filterResults: Filter.FilterResults?) {
        if (filterResults != null) {
            @Suppress("UNCHECKED_CAST")
            userNameSuggestAdapter.setSuggestionList(filterResults.values as List<UserSuggestion>)
        }
    }

    @Synchronized private fun getUsers(): List<UserSuggestion> {
        if (users == null) {
            val userlist = usersTracker.getUsers()
            users = userlist.entries.map { (key, value) -> UserSuggestion(key, value.name) }.sortedWith(userSuggestionComparator)
        }
        return users!!
    }

    @Synchronized private fun clearUserList() {
        users = null
    }

    fun shutdown() {
        usersTracker.removeUserActivityListener(listener)
    }

    private inner class UserTrackerListener : ChannelUsersTracker.UserActivityListener {
        override fun activeUserListSync() {
            clearUserList()
        }

        override fun userUpdated(uid: String) {
            clearUserList()
        }
    }
}
